### PR TITLE
More fpLookupSubdir cleanups

### PR DESCRIPTION
**Pull request by mlschroe**
This reduces the number of calls to fpLookupSubdir() and also gets rid of a poolid->str->poolid roundtrip.

### CODE REVIEW

**Comment by mlschroe** [comment]
Ok, I have to admit I'm not entirely sure what fpLookupSubdir is trying to fix. It was added by Florian in commits c6ccc90d7fef0f1b65e4bf5b77d5b800d4b53ffd and af3464a053ecb0b56cc5af494ea22955fb350757, unfortunately without a reference to some bug.

It seems to be about having a symlink in one to-be-installed package and then installing over that symlink in another to-be-installed package, but that use case does not work reliable anyway because the symlink package must be installed first for it to work.

Florian, do you remember more details what this was about?

**Comment by pmatilai** [comment]
IIRC this is the reproducer case: http://lists.rpm.org/pipermail/rpm-maint/2008-April/002051.html

**Comment by pmatilai** [comment]
In case you wonder, I remember due to encountering this during the string-pool mass changes, and I do remember having a manual test-case for commit b81b4a35240f16fa8b45156b0151fab9e130a8e8. Too bad I didn't create an automated test back then :disappointed:

**Comment by ffesti** [comment]
Yes, that's pretty much it. The finger printing code calculates a unique identifier for each file's location. This is comprised out of the device id and inode number of the parent dir and the filename. If the parent dir is not on disk yet, the closest dir is used + the path down to the parent to be installed.
fpLookupSubdir looks for symlinks that may alias the files in this - not yet installed - part of the path. Yes, it assumes that the symlink actually gets installed. But this is not an unreasonable assumption as we do a file conflict checks for this symlinks, too.

**Comment by mlschroe** [comment]
Ok, that's what I thought. But it's somewhat brittle, that example from the mail will not work if FOO-DOC is installed before FOO as then /usr/share/FOO-1 will get created as directory and the install of FOO will fail with a RPMERR_EXIST_AS_DIR error.

(I'm trying to make dir -> symlink-to-dir transactions work in the easy cases, I think this will be fixed with the changes as well.)

**Comment by mlschroe** [comment]
Regarding your b81b4a35240f16fa8b45156b0151fab9e130a8e8 commit: fpLookupSubdir's slash handling is still somewhat broken, it tends to duplicate slashes when creating the link. The fingerprint lookup fortunately calls rpmCleanPath() with gets rid of the extra slashes again.

BTW, why do the subDir entries in the fingerprints both have a leading and a trailing `/`? Is that something we should fix?

**Comment by pmatilai** [comment]
> I'm trying to make dir -> symlink-to-dir transactions work in the easy cases, I think this will be fixed with the changes as well

Wow. Good luck with that :grin: 

As for the slashes, I seem to recall looking at reducing them but something preventing that, perhaps some "hidden" dependency to rpmfi API / users expecting those slashes there and placing them back would've been impossible/hard. But it's a long time ago, things may have changed and my memory is really hazy on the details at best. 

BTW, are you expecting to push more commits to this particular PR or should we just merge it?

**Comment by mlschroe** [comment]
I didn't plan to push more commits, but now that you ask:

Does it really make sense to have the symlinks of already installed packages in the `symlinks` hash? If a symlink is deleted on disk, the code will ignore it anyway (see the "Ignore already removed (by eg %pretrans) links" comment). If it's still on disk, the fingerprint lookup code will have found it and it will not be part of the "subDir" element. So what's the point?

**Comment by pmatilai** [comment]
Good question, I've no idea. @ffesti?

**Comment by mlschroe** [comment]
I've added a commit that makes the code only consider symlinks. Florian, any objections to this?

File: lib/fprint.c [hunk 1]
```diff
@@ -459,11 +450,14 @@ void fpCachePopulate(fingerPrintCache fpc, rpmts ts, int fileCount)
...
     rpmFpHash symlinks = rpmFpHashCreate(fileCount/16+16, fpHashFunction, fpEqual, NULL, NULL);
 
+    /* create a hash of all symlinks in the new packages */
     pi = rpmtsiInit(ts);
     while ((p = rpmtsiNext(pi, 0)) != NULL) {
 	fingerPrint *fpList;
 	(void) rpmsqPoll();
 
+	if (rpmteType(p) == TR_REMOVED)
+	    continue;		/* we are only interested in new packages */
```
Comment (pmatilai) [inline lib/fprint.c]: Why not change the loop itself to only look at added packages instead, eg rpmtsiNext(pi, TR_ADDED) in the above?

Comment (mlschroe) [reply lib/fprint.c]: Oh, I didn't know about that. Force pushed.

**Comment by pmatilai** [comment]
This crashes at the "reinstall 1" test with the following (from valgrind):

> fprint.c:262:12: runtime error: member access within null pointer of type 'const struct fingerPrint'
==943715== Invalid read of size 4
==943715==    at 0x4923867: fpHashFunction (fprint.c:262)
==943715==    by 0x492777D: rpmFpHashAddEntry (rpmhash.C:168)
==943715==    by 0x4929D42: fpCachePopulate (fprint.c:515)
==943715==    by 0x49958BF: rpmtsPrepare (transaction.c:1464)
==943715==    by 0x49964CE: rpmtsRun (transaction.c:1721)
==943715==    by 0x4973989: rpmcliTransaction (rpminstall.c:291)
==943715==    by 0x4977A99: rpmInstall (rpminstall.c:632)
==943715==    by 0x10C6C9: main (rpm.c:265)
==943715==  Address 0xc is not stack'd, malloc'd or (recently) free'd

I guess the second loop will also need to be changed to look at only added packages, or something.

For an easy reproducer outside the test-suite:
> [root@lumikko-w rpm]# ./rpm -U --nodeps --ignorearch --root /srv/test tests/data/RPMS/hello-2.0-1.i686.rpm 
[root@lumikko-w rpm]# ./rpm --reinstall --nodeps --ignorearch --nodocs --root /srv/test/ tests/data/RPMS/hello-2.0-1.i686.rpm

**Comment by mlschroe** [comment]
Oh yes, we'll need to do the rpmfilesFpLookup() call for TR_REMOVED packages. Fixing...

**Comment by pmatilai** [comment]
Okay so there was more than meets the eye... thankfully caught by the test-suite.

**Comment by pmatilai** [comment]
Doh, certainly didn't intend to close but just comment and then merge.
Been multiple such mistakes from me in the last week or so, wonder if some button order or such on GH changed.

**Comment by pmatilai** [comment]
OTOH maybe I just need to adopt a merge first strategy to avoid embarrassing accidental closures...
Anyway, thanks for the nice cleanups.

**Pull request merged**